File: backend/app/services/gamification.py
"""Gamification service - port of PHP `GamificationService`.

Transparent XP / level / streak / achievement rules. Data-driven, no AI.
"""

import logging
from datetime import date, timedelta
from typing import Any

from app.repositories.quiz_repository import QuizRepository
from app.utils.supabase import SupabaseClient

logger = logging.getLogger(__name__)

LEVEL_THRESHOLDS = [0, 500, 1000, 1500, 2500, 4000, 6000, 8000, 10000, 12500, 15000]

XP_BASE = 10
XP_PER_PERCENT = 1
XP_PERFECT_BONUS = 15
XP_DAILY_BONUS = 5
XP_IMPROVE_BONUS = 15

MASTERY_IMPROVED = 60.0


def level_info(xp: int) -> dict[str, Any]:
    t = LEVEL_THRESHOLDS
    level = 1
    for i in range(len(t) - 1):
        if xp >= t[i + 1]:
            level = i + 2
        else:
            break

    last_idx = len(t) - 1
    if level > last_idx + 1:
        gap = t[last_idx] - t[last_idx - 1]
        floor = t[last_idx] + ((level - (last_idx + 1)) * gap)
    else:
        floor = t[level - 1]
        nxt = t[level] if level < len(t) else (t[last_idx] + (t[last_idx] - t[last_idx - 1]))
        gap = max(1, nxt - floor)

    xp_into = max(0, xp - floor)
    xp_needed = max(1, gap)
    return {
        "level": level,
        "xp_into": xp_into,
        "xp_for_level": floor,
        "xp_next": max(0, xp_needed - xp_into),
        "progress": min(1.0, xp_into / xp_needed) if xp_needed > 0 else 1.0,
    }


def compute_attempt_xp(score: int, total: int, num_improved: int, streak_increased: bool) -> dict:
    percent = round((score / total) * 100) if total > 0 else 0
    breakdown = {"base": XP_BASE, "accuracy": int(percent)}
    if total > 0 and percent >= 100:
        breakdown["perfect"] = XP_PERFECT_BONUS
    if streak_increased:
        breakdown["daily"] = XP_DAILY_BONUS
    if num_improved > 0:
        breakdown["improvement"] = num_improved * XP_IMPROVE_BONUS
    return {"total": int(sum(breakdown.values())), "breakdown": breakdown}


def next_streak(last_active: str | None) -> dict:
    today = date.today()
    if last_active == today.isoformat():
        return {"streak": 0, "increased": False}
    return {"streak": 1, "increased": True}


class GamificationService:
    def __init__(self, db: SupabaseClient, repo: QuizRepository):
        self._db = db
        self._repo = repo

    async def award_attempt(
        self,
        user_id: str,
        token: str,
        attempt: dict,
        profile: dict,
        attempt_count: int,
        concept_before: list[dict],
    ) -> dict[str, Any]:
        current_xp = int(profile.get("xp") or 0)
        current_level = int(profile.get("level") or 1)
        current_streak = int(profile.get("streak") or 0)
        score = int(attempt.get("score") or 0)
        total = int(attempt.get("total") or 0)
        percent = float(attempt.get("percent") or 0)

        streak_info = next_streak(str(profile["last_active"]) if profile.get("last_active") else None)
        new_streak = current_streak + 1 if streak_info["increased"] else current_streak

        num_improved = await self._count_improved_concepts(user_id, token, concept_before)

        xp = compute_attempt_xp(score, total, num_improved, streak_info["increased"])
        new_xp = current_xp + xp["total"]
        info = level_info(new_xp)
        leveled_up = info["level"] > current_level

        try:
            await self._db.update(
                "profiles",
                {
                    "xp": new_xp,
                    "level": info["level"],
                    "streak": new_streak,
                    "last_active": date.today().isoformat(),
                },
                {"id": f"eq.{user_id}"},
                token=token,
            )
        except Exception as exc:
            logger.warning("Could not persist gamification update for %s: %s", user_id, exc)

        state = {
            "attempt_count": attempt_count,
            "this_percent": percent,
            "score": score,
            "total": total,
            "streak": new_streak,
            "num_improved": num_improved,
        }
        new_achievements = await self.grant_eligible(user_id, token, state)

        return {
            "xp_earned": xp["total"],
            "xp_breakdown": xp["breakdown"],
            "level": info["level"],
            "leveled_up": leveled_up,
            "level_info": info,
            "streak": new_streak,
            "streak_increased": streak_info["increased"],
            "num_improved": num_improved,
            "new_achievements": new_achievements,
        }

    async def _count_improved_concepts(
        self, user_id: str, token: str, concept_before: list[dict]
    ) -> int:
        try:
            after = await self._repo.list_concept_performance(user_id, token)
        except Exception:
            return 0
        if not concept_before:
            return 0

        before = {
            str(c.get("concept") or "General"): {
                "attempts": int(c.get("attempts") or 0),
                "mastery": float(c.get("mastery") or 0),
            }
            for c in concept_before
        }

        improved = 0
        for c in after:
            name = str(c.get("concept") or "General")
            prev = before.get(name)
            if prev is None:
                continue
            was_weak = prev["mastery"] < MASTERY_IMPROVED
            now_strong = float(c.get("mastery") or 0) >= MASTERY_IMPROVED
            touched = int(c.get("attempts") or 0) > prev["attempts"]
            if was_weak and now_strong and touched:
                improved += 1
        return improved

    async def _catalogue(self, token: str, order: str | None = None) -> list[dict]:
        try:
            return await self._db.select(
                "achievements",
                columns="id, code, name, icon, description",
                **({"order": order} if order else {}),
                token=token,
            )
        except Exception:
            return []

    async def _owned_codes(self, user_id: str, token: str, catalogue: list[dict]) -> dict[str, bool]:
        id_to_code = {str(a.get("id") or ""): str(a.get("code") or "") for a in catalogue}
        owned: dict[str, bool] = {}
        try:
            granted = await self._db.select(
                "user_achievements",
                columns="achievement_id",
                filters={"user_id": f"eq.{user_id}"},
                token=token,
            )
        except Exception:
            return owned
        for g in granted:
            code = id_to_code.get(str(g.get("achievement_id") or ""))
            if code:
                owned[code] = True
        return owned

    async def grant_eligible(self, user_id: str, token: str, state: dict) -> list[dict]:
        catalogue = await self._catalogue(token)
        owned = await self._owned_codes(user_id, token, catalogue)

        new_ones = []
        for a in catalogue:
            code = str(a.get("code") or "")
            if code == "" or code in owned:
                continue
            if not self._is_met(code, state):
                continue
            try:
                await self._db.insert(
                    "user_achievements",
                    {"user_id": user_id, "achievement_id": a.get("id")},
                    token=token,
                )
            except Exception:
                continue
            new_ones.append(
                {
                    "code": code,
                    "name": str(a.get("name") or code),
                    "icon": str(a.get("icon") or "bi-trophy"),
                    "description": str(a.get("description") or ""),
                }
            )
        return new_ones

    async def grant_by_code(self, user_id: str, token: str, code: str) -> dict | None:
        catalogue = await self._catalogue(token)
        for a in catalogue:
            if str(a.get("code") or "") != code:
                continue
            try:
                await self._db.insert(
                    "user_achievements",
                    {"user_id": user_id, "achievement_id": a.get("id")},
                    token=token,
                )
            except Exception:
                return None
            return {
                "code": code,
                "name": str(a.get("name") or code),
                "icon": str(a.get("icon") or "bi-trophy"),
                "description": str(a.get("description") or ""),
            }
        return None

    async def catalogue_with_state(self, user_id: str, token: str) -> dict:
        all_rows = await self._catalogue(token, order="created_at.asc")
        if not all_rows:
            return {"all": [], "unlocked": []}
        owned = await self._owned_codes(user_id, token, all_rows)
        unlocked = []
        for a in all_rows:
            a = dict(a)
            a["unlocked"] = bool(owned.get(str(a.get("code") or "")))
            if a["unlocked"]:
                unlocked.append(a)
        return {"all": all_rows, "unlocked": unlocked}

    @staticmethod
    def _is_met(code: str, state: dict) -> bool:
        if code == "first_quiz":
            return int(state.get("attempt_count") or 0) >= 1
        if code == "quiz_master":
            return int(state.get("attempt_count") or 0) >= 10
        if code == "percent_90":
            return float(state.get("this_percent") or 0) >= 90
        if code == "perfect_score":
            return int(state.get("score") or 0) > 0 and float(state.get("this_percent") or 0) >= 100
        if code == "streak_7":
            return int(state.get("streak") or 0) >= 7
        if code == "improvement_champion":
            return int(state.get("num_improved") or 0) >= 1
        return False

File: includes/Gamification.php
<?php
/**
 * QuizSphere - Gamification service
 * -------------------------------------------------------------
 * Transparent XP, level, streak and achievement rules for the platform.
 *
 * All rules are deliberately simple and data-driven so they are easy to
 * reason about, test and modify. AI is never involved here.
 *
 * LEVELS
 *   XP thresholds are defined in LEVEL_THRESHOLDS. The user's level is the
 *   highest index whose threshold is met (first entry is Level 1 at 0 XP).
 *   When XP passes the final threshold, the last gap is reused to keep
 *   levelling smooth at high values.
 *
 * XP PER COMPLETED QUIZ (each completed attempt, awarded once)
 *   + 10          base completion
 *   + 1 per %     accuracy (rounded), 0..100
 *   + 15          perfect score bonus (100%)
 *   + 5           first attempt of the day (streak day)
 *   + 15 per      concept improved from weak (<60%) to >=60% this attempt
 *
 * STREAK
 *   Consecutive days with at least one completed attempt. Repeats on the
 *   same day do not reset the streak. Inactivity resets it to 1.
 *
 * ACHIEVEMENTS
 *   Granted only when the actual criteria are met (see the catalogue seeds in
 *   database/migrations/phase6.sql). Newly-unlocked badges are returned so
 *   the UI can celebrate them.
 */

declare(strict_types=1);

class GamificationService
{
    /** XP needed to reach each level (index 0 => Level 1 at 0 XP). */
    public const LEVEL_THRESHOLDS = [0, 500, 1000, 1500, 2500, 4000, 6000, 8000, 10000, 12500, 15000];

    // XP rewards (kept together so the rules are obvious).
    public const XP_BASE          = 10;
    public const XP_PER_PERCENT   = 1;
    public const XP_PERFECT_BONUS = 15;
    public const XP_DAILY_BONUS   = 5;
    public const XP_IMPROVE_BONUS = 15;

    public const MASTERY_IMPROVED = 60; // >=60% = no longer weak

    private SupabaseClient $db;
    private QuizRepository $repo;

    public function __construct(SupabaseClient $db, QuizRepository $repo)
    {
        $this->db = $db;
        $this->repo = $repo;
    }

    /**
     * Calculate the level for a given XP total.
     *
     * @return array{level:int, xp_into:int, xp_for_level:int, xp_next:int, progress:float}
     *               xp_into = XP earned within the current level,
     *               xp_for_level = total XP required to reach current level,
     *               xp_next = remaining XP to the next level,
     *               progress = 0..1 within the current level.
     */
    public function levelInfo(int $xp): array
    {
        $t = self::LEVEL_THRESHOLDS;
        $level = 1;
        for ($i = 0; $i < count($t) - 1; $i++) {
            if ($xp >= $t[$i + 1]) {
                $level = $i + 2;
            } else {
                break;
            }
        }

        $lastIdx = count($t) - 1;
        $gap = 0;
        if ($level > $lastIdx + 1) {
            // Beyond the defined table: reuse the final gap.
            $gap = $t[$lastIdx] - $t[$lastIdx - 1];
            $floor = $t[$lastIdx] + (($level - ($lastIdx + 1)) * $gap);
        } else {
            $floor = $t[$level - 1];
            $next = $t[$level] ?? ($t[$lastIdx] + ($t[$lastIdx] - $t[$lastIdx - 1]));
            $gap = max(1, $next - $floor);
        }

        $xpInto = max(0, $xp - $floor);
        $xpNeeded = max(1, $gap);

        return [
            'level'        => $level,
            'xp_into'      => $xpInto,
            'xp_for_level' => $floor,
            'xp_next'      => max(0, $xpNeeded - $xpInto),
            'progress'     => $xpNeeded > 0 ? min(1.0, $xpInto / $xpNeeded) : 1.0,
        ];
    }

    /**
     * Transparent XP calculation for a completed attempt.
     *
     * @return array{total:int, breakdown:array<string,int>}
     */
    public function computeAttemptXp(int $score, int $total, int $numImproved, int $newStreak, bool $streakIncreased): array
    {
        $percent = $total > 0 ? round(($score / $total) * 100) : 0;
        $breakdown = [];
        $breakdown['base'] = self::XP_BASE;
        $breakdown['accuracy'] = (int) $percent;
        if ($total > 0 && $percent >= 100) {
            $breakdown['perfect'] = self::XP_PERFECT_BONUS;
        }
        if ($streakIncreased) {
            $breakdown['daily'] = self::XP_DAILY_BONUS;
        }
        if ($numImproved > 0) {
            $breakdown['improvement'] = $numImproved * self::XP_IMPROVE_BONUS;
        }

        return ['total' => (int) array_sum($breakdown), 'breakdown' => $breakdown];
    }

    /**
     * Compute the next streak given the profile's last_active date.
     *
     * @param string|null $lastActive Y-m-d or null.
     * @return array{streak:int, increased:bool}
     */
    public function nextStreak(?string $lastActive): array
    {
        $today = date('Y-m-d');
        $yesterday = date('Y-m-d', strtotime('-1 day'));

        if ($lastActive === $today) {
            return ['streak' => 0, 'increased' => false]; // streak stays as-is (same day)
        }
        if ($lastActive === $yesterday) {
            return ['streak' => 1, 'increased' => true];
        }
        // New streak (first attempt or after a gap).
        return ['streak' => 1, 'increased' => true];
    }

    /**
     * Main reward flow for a completed, persisted attempt.
     *
     * @param array $attempt      {score:int, total:int, percent:float}
     * @param array $profile      Current profiles row (xp, level, streak, last_active).
     * @param int   $attemptCount Total completed attempts, this one included.
     * @param array $conceptBefore Concept performance rows BEFORE this attempt
     *                            (from listConceptPerformance) for improvement detection.
     * @return array{xp_earned:int, xp_breakdown:array, level:int, leveled_up:bool,
     *               level_info:array, streak:int, streak_increased:bool,
     *               new_achievements:array}
     */
    public function awardAttempt(string $userId, string $token, array $attempt, array $profile, int $attemptCount, array $conceptBefore): array
    {
        $currentXp = (int) ($profile['xp'] ?? 0);
        $currentLevel = (int) ($profile['level'] ?? 1);
        $currentStreak = (int) ($profile['streak'] ?? 0);
        $score = (int) ($attempt['score'] ?? 0);
        $total = (int) ($attempt['total'] ?? 0);
        $percent = (float) ($attempt['percent'] ?? 0);

        // 1. Streak.
        $streakInfo = $this->nextStreak(isset($profile['last_active']) ? (string) $profile['last_active'] : null);
        // Same-day repeat: keep the current streak (no reset, no bonus).
        $newStreak = $streakInfo['increased'] ? $currentStreak + 1 : $currentStreak;

        // 2. Weak-concept improvement this attempt.
        $numImproved = $this->countImprovedConcepts($userId, $token, $conceptBefore);

        // 3. XP.
        $xp = $this->computeAttemptXp($score, $total, $numImproved, $newStreak, $streakInfo['increased']);
        $newXp = $currentXp + $xp['total'];
        $levelInfo = $this->levelInfo($newXp);
        $leveledUp = $levelInfo['level'] > $currentLevel;

        // 4. Persist profile updates (RLS lets the user update their own row).
        try {
            $this->db->update('profiles', [
                'xp'          => $newXp,
                'level'       => $levelInfo['level'],
                'streak'      => $newStreak,
                'last_active' => date('Y-m-d'),
            ], ['id' => "eq.$userId"], $token);
        } catch (SupabaseException $e) {
            AppLogger::warning('Could not persist gamification update', ['user_id' => $userId, 'message' => $e->getMessage()], 'gamification');
        }

        // 5. Achievements.
        $state = [
            'attempt_count' => $attemptCount,
            'this_percent'  => $percent,
            'score'         => $score,
            'total'         => $total,
            'streak'        => $newStreak,
            'num_improved'  => $numImproved,
        ];
        $newAchievements = $this->grantEligible($userId, $token, $state);

        return [
            'xp_earned'        => $xp['total'],
            'xp_breakdown'     => $xp['breakdown'],
            'level'            => $levelInfo['level'],
            'leveled_up'       => $leveledUp,
            'level_info'       => $levelInfo,
            'streak'           => $newStreak,
            'streak_increased' => $streakInfo['increased'],
            'num_improved'     => $numImproved,
            'new_achievements' => $newAchievements,
        ];
    }

    /**
     * Count how many previously-weak concepts the user improved in this
     * attempt by comparing cumulative mastery before vs after.
     */
    private function countImprovedConcepts(string $userId, string $token, array $conceptBefore): int
    {
        try {
            $after = $this->repo->listConceptPerformance($userId, $token);
        } catch (SupabaseException $e) {
            return 0;
        }

        if ($conceptBefore === []) {
            return 0;
        }

        $beforeByConcept = [];
        foreach ($conceptBefore as $c) {
            $name = (string) ($c['concept'] ?? 'General');
            $beforeByConcept[$name] = [
                'attempts' => (int) ($c['attempts'] ?? 0),
                'mastery'  => (float) ($c['mastery'] ?? 0),
            ];
        }

        $improved = 0;
        foreach ($after as $c) {
            $name = (string) ($c['concept'] ?? 'General');
            $prev = $beforeByConcept[$name] ?? null;
            if ($prev === null) {
                continue; // brand-new concept, not an "improvement"
            }
            $wasWeak = $prev['mastery'] < self::MASTERY_IMPROVED;
            $nowStrong = (float) ($c['mastery'] ?? 0) >= self::MASTERY_IMPROVED;
            $touched = (int) ($c['attempts'] ?? 0) > $prev['attempts'];
            if ($wasWeak && $nowStrong && $touched) {
                $improved++;
            }
        }
        return $improved;
    }

    /**
     * Evaluate and grant every achievement whose criteria are now met.
     *
     * @return array[] Newly granted achievements [{code,name,icon,description}]
     */
    public function grantEligible(string $userId, string $token, array $state): array
    {
        $catalogue = [];
        try {
            $catalogue = $this->db->select('achievements', ['columns' => 'id, code, name, icon, description'], $token);
        } catch (SupabaseException $e) {
            return [];
        }

        // Map achievement id -> code from the catalogue.
        $idToCode = [];
        foreach ($catalogue as $a) {
            $idToCode[(string) ($a['id'] ?? '')] = (string) ($a['code'] ?? '');
        }

        // Already-owned codes (own rows via RLS).
        $owned = [];
        try {
            $granted = $this->db->select('user_achievements', [
                'columns' => 'achievement_id',
                'filter'  => ['user_id' => "eq.$userId"],
            ], $token);
            foreach ($granted as $g) {
                $code = $idToCode[(string) ($g['achievement_id'] ?? '')] ?? null;
                if (is_string($code) && $code !== '') {
                    $owned[$code] = true;
                }
            }
        } catch (SupabaseException $e) {
            // proceed empty
        }

        $newOnes = [];
        foreach ($catalogue as $a) {
            $code = (string) ($a['code'] ?? '');
            if ($code === '' || isset($owned[$code])) {
                continue;
            }
            if (!$this->isMet($code, $state)) {
                continue;
            }
            // Grant.
            try {
                $this->db->insert('user_achievements', [
                    'user_id'        => $userId,
                    'achievement_id' => $a['id'],
                ], $token);
            } catch (SupabaseException $e) {
                continue; // already granted concurrently; skip
            }
            $newOnes[] = [
                'code'        => $code,
                'name'        => (string) ($a['name'] ?? $code),
                'icon'        => (string) ($a['icon'] ?? 'bi-trophy'),
                'description' => (string) ($a['description'] ?? ''),
            ];
        }
        return $newOnes;
    }

    /**
     * Grant a single achievement by code (used for non-attempt events such as
     * earning a certificate). Returns the achievement row or null.
     */
    public function grantByCode(string $userId, string $token, string $code): ?array
    {
        $catalogue = [];
        try {
            $catalogue = $this->db->select('achievements', ['columns' => 'id, code, name, icon, description'], $token);
        } catch (SupabaseException $e) {
            return null;
        }
        foreach ($catalogue as $a) {
            if ((string) ($a['code'] ?? '') !== $code) {
                continue;
            }
            try {
                $this->db->insert('user_achievements', [
                    'user_id'        => $userId,
                    'achievement_id' => $a['id'],
                ], $token);
            } catch (SupabaseException $e) {
                return null; // already owned
            }
            return [
                'code'        => $code,
                'name'        => (string) ($a['name'] ?? $code),
                'icon'        => (string) ($a['icon'] ?? 'bi-trophy'),
                'description' => (string) ($a['description'] ?? ''),
            ];
        }
        return null;
    }

    /**
     * Return all achievements with the user's unlock state, for the UI.
     *
     * @return array{all: array[], unlocked: array[]}
     */
    public function catalogueWithState(string $userId, string $token): array
    {
        $all = [];
        try {
            $all = $this->db->select('achievements', ['columns' => 'id, code, name, icon, description', 'order' => 'created_at.asc'], $token);
        } catch (SupabaseException $e) {
            return ['all' => [], 'unlocked' => []];
        }

        $idToCode = [];
        foreach ($all as $a) {
            $idToCode[(string) ($a['id'] ?? '')] = (string) ($a['code'] ?? '');
        }

        $owned = [];
        try {
            $granted = $this->db->select('user_achievements', [
                'columns' => 'achievement_id',
                'filter'  => ['user_id' => "eq.$userId"],
            ], $token);
            foreach ($granted as $g) {
                $code = $idToCode[(string) ($g['achievement_id'] ?? '')] ?? null;
                if (is_string($code) && $code !== '') {
                    $owned[$code] = true;
                }
            }
        } catch (SupabaseException $e) {
            // ignore
        }

        $unlocked = [];
        foreach ($all as $a) {
            $a['unlocked'] = isset($owned[(string) ($a['code'] ?? '')]);
            if ($a['unlocked']) {
                $unlocked[] = $a;
            }
        }
        return ['all' => $all, 'unlocked' => $unlocked];
    }

    /** Evaluate a single achievement code against the current state. */
    private function isMet(string $code, array $state): bool
    {
        switch ($code) {
            case 'first_quiz':
                return (int) ($state['attempt_count'] ?? 0) >= 1;
            case 'quiz_master':
                return (int) ($state['attempt_count'] ?? 0) >= 10;
            case 'percent_90':
                return (float) ($state['this_percent'] ?? 0) >= 90;
            case 'perfect_score':
                return (int) ($state['score'] ?? 0) > 0 && (float) ($state['this_percent'] ?? 0) >= 100;
            case 'streak_7':
                return (int) ($state['streak'] ?? 0) >= 7;
            case 'improvement_champion':
                return (int) ($state['num_improved'] ?? 0) >= 1;
            default:
                return false;
        }
    }
}
